pkg/controller/pinned-image-set/pinned_image_set_controller.go

```diff
@@ -0,0 +1,698 @@
+			Reason:             ErrorReason,
+			Message:            err.Error(),
+			LastTransitionTime: metav1.NewTime(time.Now()),
+		})
+		if statusErr != nil {
+			klog.Warningf("Error updating status for PinnedImageSet %v: %v", key, statusErr)
+		}
+		return err
+	}
+	statusErr := ctrl.updateStatus(imageSet, metav1.Condition{
+		Type:               configDegradedCondition,
+		Status:             metav1.ConditionFalse,
+		Reason:             ConfigDeployedReason,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+	})
+	if statusErr != nil {
+		klog.Warningf("Error updating status for PinnedImageSet %v: %v", key, statusErr)
+	}
+
+	// report prefetch image status
+	err = ctrl.ensureNodeImagePrefetch(imageSet)
+	if err != nil {
+		statusErr := ctrl.updateStatus(imageSet, metav1.Condition{
+			Type:               prefetchDegradedCondition,
+			Status:             metav1.ConditionTrue,
+			Reason:             ErrorReason,
+			Message:            err.Error(),
+			LastTransitionTime: metav1.NewTime(time.Now()),
+		})
+		if statusErr != nil {
+			klog.Warningf("Error updating status for PinnedImageSet %v: %v", key, statusErr)
+		}
+		return err
+	}
+	statusErr = ctrl.updateStatus(imageSet, metav1.Condition{
+		Type:               prefetchDegradedCondition,
+		Status:             metav1.ConditionFalse,
+		Reason:             PrefetchCompletedReason,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+	})
+	if statusErr != nil {
+		klog.Warningf("Error updating status for PinnedImageSet %v: %v", key, statusErr)
+	}
+
+	return nil
+}
+
+func (ctrl *Controller) ensurePinnedImageSet(imageSet *mcfgv1.PinnedImageSet) error {
+	// Check for Deleted PinnedImageSet and optionally delete finalizers
+	if imageSet.DeletionTimestamp != nil {
+		if len(imageSet.GetFinalizers()) > 0 {
+			klog.V(2).Infof("sync() deleting finalizers for PinnedImageSet: %s\n", imageSet.Name)
+			return ctrl.cascadeDelete(imageSet)
+		}
+		return nil
+	}
+
+	// get pools that match node selector
+	mcpPools, err := ctrlcommon.GetPoolsForPinnedImageSet(ctrl.mcpLister, imageSet)
+	if err != nil {
+		return err
+	}
+	if len(mcpPools) == 0 {
+		return fmt.Errorf("%w : %s", ErrNoMatchingPool, imageSet.Name)
+	}
+
+	for _, pool := range mcpPools {
+		managedKey, err := getMachineConfigKey(pool, ctrl.mcfgClient, imageSet)
+		if err != nil {
+			return err
+		}
+
+		mc, err := ctrl.mcfgClient.MachineconfigurationV1().MachineConfigs().Get(context.TODO(), managedKey, metav1.GetOptions{})
+		// check if not found error
+		isNotFound := apierrors.IsNotFound(err)
+		if err != nil && !isNotFound {
+			return err
+		}
+		// If we have seen this generation and the sync didn't fail, then skip
+		if !isNotFound && imageSet.Status.ObservedGeneration >= imageSet.Generation {
+			// But we still need to compare the generated controller version because during an upgrade we need a new one
+			mcCtrlVersion := mc.Annotations[ctrlcommon.GeneratedByControllerVersionAnnotationKey]
```

Comment:
do you just want to throw this in the above if? `if !isNotFound && imageSet.Status.ObservedGeneration >= imageSet.Generation && mc.Annotations[ctrlcommon.GeneratedByControllerVersionAnnotationKey] != version.hash`? or do you want to really return nil here?